improve instrumentation doc (#6777)
DataDog/dd-trace-java: docs/how_instrumentations_work.md

retrieve a value if you use the exact same key object as when it was set. Using a different object that is “`.equals()`”
to the first will yield nothing.

> Also, the ContextStore is not supporting null keys, hence null check must be enforced before using an object as a key.

Reviewer: ```suggestion
Since `ContextStore` does not support null keys, null checks must be enforced _before_ using an object as a key.
```

or [reset](https://github.com/DataDog/dd-trace-java/blob/9d5c7ea524cfec982176e687a489fc8c2865e445/dd-java-agent/instrumentation/java-http-client/src/main/java11/datadog/trace/instrumentation/httpclient/SendAdvice.java#L44))
when exiting.

> Please note that it works only if the methods are called on the same thread (since the counter is a threadlocal variable)

Reviewer: ```suggestion
This only works if the methods are called on the same thread since the counter is a ThreadLocal variable.
```
